cmd/kubelet/app/server.go

```diff
@@ -186,6 +186,120 @@ func Run(s *options.KubeletServer, kubeDeps *kubelet.Dependencies) error {
 	return nil
 }
 
+// setKubeDepsClients initializes clients if not standalone mode and any of the clients are not provided
+func setKubeDepsClients(s *options.KubeletServer, kubeDeps *kubelet.Dependencies, nodeName types.NodeName) error {
+	clientConfig, err := CreateAPIServerClientConfig(s)
+	if err == nil {
+		clientCertificateManager, err := makeKubeletClientCertificateManager(s, clientConfig, nodeName)
+		if err != nil {
+			return err
+		}
+
+		kubeClient, err := makeKubeDepsKubeClient(kubeDeps, clientConfig)
+		if err != nil {
+			startClientCertificateManager(kubeClient, clientCertificateManager)
+		}
+
+		setKubeDepsExternalKubeClient(kubeDeps, clientConfig)
+		setKubeDepsEventClient(s, kubeDeps, clientConfig)
+		setKubeDepsHeartbeatClient(s, kubeDeps, clientConfig)
+
+	} else {
+		switch {
+		case s.RequireKubeConfig:
+			return fmt.Errorf("invalid kubeconfig: %v", err)
+		case s.KubeConfig.Provided():
+			glog.Warningf("invalid kubeconfig: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func makeKubeletClientCertificateManager(
+	s *options.KubeletServer,
+	clientConfig *restclient.Config,
+	nodeName types.NodeName) (certificate.Manager, error) {
+
+	var clientCertificateManager certificate.Manager
+	var err error
+	if s.RotateCertificates && utilfeature.DefaultFeatureGate.Enabled(features.RotateKubeletClientCertificate) {
```

Comment:
Do these checks outside this function, I'd expect a function named `makeKubeletClientCertificateManager` to only be called once you know you need to make one.

Reply:
ok